File: src/java/magmac/app/stage/InlinePassResult.java
package magmac.app.stage;

import magmac.api.None;
import magmac.api.Option;
import magmac.api.Some;
import magmac.api.result.Ok;
import magmac.app.compile.error.CompileResult;
import magmac.app.compile.error.CompileResults;
import magmac.app.compile.node.Node;
import magmac.app.stage.parse.ParseState;

import java.util.function.Supplier;

public record InlinePassResult(Option<CompileResult<ParseUnit>> option) implements ParseResult {
    public static ParseResult empty() {
        return new InlinePassResult(new None<>());
    }

    public static ParseResult from(ParseState state, Node node) {
        return new InlinePassResult(new Some<>(CompileResults.fromOk(new ParseUnit(state, node))));
    }

    @Override
    public CompileResult<ParseUnit> orElseGet(Supplier<ParseUnit> other) {
        return this.option.orElseGet(() -> CompileResults.fromResult(new Ok<>(other.get())));
    }
}

File: src/web/magmac/app/stage/InlinePassResult.ts
import { None } from "../../../magmac/api/None";
import { Option } from "../../../magmac/api/Option";
import { Tuple2 } from "../../../magmac/api/Tuple2";
import { Ok } from "../../../magmac/api/result/Ok";
import { CompileResult } from "../../../magmac/app/compile/error/CompileResult";
import { InlineCompileResult } from "../../../magmac/app/compile/error/InlineCompileResult";
import { Node } from "../../../magmac/app/compile/node/Node";
import { ParseState } from "../../../magmac/app/stage/parse/ParseState";
import { Supplier } from "../../../java/util/function/Supplier";
export class InlinePassResult {
	empty() : PassResult {return new InlinePassResult( new None<>( ));}
	orElseGet(other : Supplier<Tuple2<ParseState, Node>>) : CompileResult<Tuple2<ParseState, Node>> {return this.option.orElseGet( () ->InlineCompileResult.fromResult( new Ok<>( other.get( ))));}
}
